src/kalai/pass/kalai/a_annotate_ast.clj

```diff
@@ -49,42 +52,65 @@
     #'ref
     #'agent})
 
-(defn ast-t [ast]
-  (m/rewrite ast
-    ;; (atom x)
-    {:op   :invoke
-     :fn   {:var (m/pred ref-vars)}
-     :args [?value . _ ...]}
-    ~(ast-t ?value)
-
-    ;; ^{:t :long, :tag Long} x
-    {:op   :with-meta
-     :meta {:form {:t ?t :tag ?tag}}
-     :expr ?expr}
-    ~(or ?t
-         (get types/java-types ?tag)
-         (ast-t ?expr))
-
-    {:op   :local
-     :form ?form}
-    ~(resolve-t ?form ast)
-
-    ;; Last resort: 1
-    {:op  :const
-     :val ?val}
-    ~(get types/java-types (type ?val))
-
-    ;; Last resort: Clojure type inferred
-    {:o-tag (m/pred some? ?o-tag)}
-    ~(or (get types/java-types ?o-tag)
-         ;; TODO: breaking the rules for interop...
-         ;; is this a bad idea?
-         ;; note that some o-tags are unhelpful,
-         ;; like PersistentMap
-         (#{StringBuffer} ?o-tag))
-
-    ?else
-    nil))
+(defn ast-t
+  "Return the type represented by an AST node"
+  ([ast]
+   (ast-t ast ast))
+  ([ast root]
+   (m/rewrite ast
+     ;; (atom x)
+     {:op   :invoke
+      :fn   {:var (m/pred ref-vars)}
+      :args [?value . _ ...]}
+     ~(ast-t ?value ast)
+
+     ;; ^{:t :long, :tag Long} x
+     {:op   :with-meta
+      :meta {:form {:t ?t :tag ?tag}}
+      :expr ?expr}
+     ~(or ?t
+          (get types/java-types ?tag)
+          (ast-t ?expr ast))
+
+     ;; Following the type inference precedence order, if we don't see type
```

Comment:
This comment is the correct logic. It is a bit confusing right now because the clause it prefixes is the first part of the precedence (a binding that has a type which we were able to figure out via `resolve-t` which checked the t and tag). The next clause is the one that looks at the initial value.

Reply:
I think I understand... See if the new verbiage in the comment gets it right?

Reply:
💯  😄